src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilteringFirstPass.java

```diff
@@ -0,0 +1,207 @@
+        filterStats = new HashMap<>();
+        readyForSecondPass = false;
+        this.tumorSample = tumorSample;
+    }
+
+    public boolean isReadyForSecondPass() { return readyForSecondPass; }
+
+    public FilterStats getFilterStats(final String filterName){
+        Utils.validateArg(filterStats.containsKey(filterName), "invalid filter name: " + filterName);
+        return filterStats.get(filterName);
+    }
+
+    public boolean isOnFilteredHaplotype(final VariantContext vc, final int maxDistance) {
+
+        final Genotype tumorGenotype = vc.getGenotype(tumorSample);
+
+        if (!hasPhaseInfo(tumorGenotype)) {
+            return false;
+        }
+
+        final String pgt = (String) tumorGenotype.getExtendedAttribute(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_GT_KEY, "");
+        final String pid = (String) tumorGenotype.getExtendedAttribute(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_ID_KEY, "");
+        final int position = vc.getStart();
+
+        final Pair<String, Integer> filteredCall = filteredPhasedCalls.get(pid);
+        if (filteredCall == null) {
+            return false;
+        }
+
+        // Check that vc occurs on the filtered haplotype
+        return filteredCall.getLeft().equals(pgt) && Math.abs(filteredCall.getRight() - position) <= maxDistance;
+    }
+
+    public void add(final FilterResult filterResult, final VariantContext vc) {
+        filterResults.add(filterResult);
+        final Genotype tumorGenotype = vc.getGenotype(tumorSample);
+
+        if (!filterResult.getFilters().isEmpty() && hasPhaseInfo(tumorGenotype)) {
+            final String pgt = (String) tumorGenotype.getExtendedAttribute(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_GT_KEY, "");
+            final String pid = (String) tumorGenotype.getExtendedAttribute(GATKVCFConstants.HAPLOTYPE_CALLER_PHASING_ID_KEY, "");
+            final int position = vc.getStart();
+            filteredPhasedCalls.put(pid, new ImmutablePair<>(pgt, position));
+        }
+    }
+
+    public void learnModelForSecondPass(final double requestedFPR){
+        final double[] readOrientationPosteriors = getFilterResults().stream()
+                .filter(r -> r.getFilters().isEmpty())
+                .mapToDouble(r -> r.getReadOrientationPosterior())
+                .toArray();
+
+        final FilterStats readOrientationFilterStats = calculateThresholdForReadOrientationFilter(readOrientationPosteriors, requestedFPR);
+        filterStats.put(GATKVCFConstants.READ_ORIENTATION_ARTIFACT_FILTER_NAME, readOrientationFilterStats);
+        readyForSecondPass = true;
+    }
+
+    /**
+     *
+     * Compute the filtering threshold that ensures that the false positive rate among the resulting pass variants
+     * will not exceed the requested false positive rate
+     *
+     * @param posteriors A list of posterior probabilities, which gets sorted
+     * @param requestedFPR We set the filtering threshold such that the FPR doesn't exceed this value
```

Comment:
Since this is a public static method, please document and have checks for requestedFPR.  For example, can it be negative?

Reply:
done